File: 140.R
# Modified Fibonacci golden nuggets

suppressPackageStartupMessages(library(Rmpfr))

options(scipen=100)

Args    <- commandArgs()
ordinal <- ifelse(is.na(Args[6]), 30, as.numeric(Args[6]))

nuggets <- vector()
n <- 0
up_or_down <- 1
while ( length(nuggets) < ordinal ) {
    n <- n + up_or_down
    discriminant <- sqrt(5 * n * n + 14 * n + 1)
    if ( floor(discriminant) == discriminant ) {
        nuggets <- c(nuggets, n)
        num_nuggets <- length(nuggets)
        if ( num_nuggets > 2 ) {
            up_or_down <- -1
            n <- floor(nuggets[num_nuggets - 1] *
                       nuggets[num_nuggets] / nuggets[num_nuggets - 2])
        }
    }
}

cat(sum(nuggets), fill=TRUE)

File: 140.js
#!/usr/bin/env node

// Modified Fibonacci golden nuggets

var program = require('commander');

program
    .version('0.1.0')
    .description('Modified Fibonacci golden nuggets')
    .option('-o, --ordinal <int>',
        'The ordinal of the last golden nugget', Number, 30)
    .parse(process.argv);

var sum = require('compute-sum');

var nuggets = [];
var n = 0;
var up_or_down = 1;
while (nuggets.length < program.ordinal) {
    n += up_or_down;
    var discriminant = Math.sqrt(5 * n * n + 14 * n + 1);
    if (Math.floor(discriminant) == discriminant) {
        nuggets.push(n);
        if (nuggets.length > 2) {
            up_or_down = -1;
            n = Math.floor(nuggets[nuggets.length - 2] *
                nuggets[nuggets.length - 1] / nuggets[nuggets.length - 3]);
        }
    }
}

console.log(sum(nuggets));
